services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java

```diff
@@ -1,52 +1,303 @@
+    radiuses.add(null);
+
+    RouteOptions routeOptions = routeOptions()
+        .toBuilder()
+        .radiusesList(radiuses)
+        .build();
+
+    assertEquals(radiuses, routeOptions.radiusesList());
+    assertEquals(";;5.7;;4.4;9.9", routeOptions.radiuses());
+  }
+
+  @Test
+  public void bearingsString() {
+    String bearingsString = ";5.1,7.4;;";
+
+    RouteOptions routeOptions = routeOptions()
+        .toBuilder()
+        .bearings(bearingsString)
+        .build();
+
+    List<Double> bearing = new ArrayList<>();
+    bearing.add(5.1);
+    bearing.add(7.4);
+
+    assertEquals(";5.1,7.4", routeOptions.bearings());
+
+    List<List<Double>> bearings = routeOptions.bearingsList();
+    assertEquals(2, bearings.size());
+    assertEquals(null, bearings.get(0));
+    assertEquals(bearing, bearings.get(1));
+  }
+
+  @Test
+  public void bearingsList() {
+    List<Double> bearing1 = new ArrayList<>();
+    bearing1.add(1.1);
+    bearing1.add(2.2);
+
+    List<Double> bearing2 = new ArrayList<>();
+    bearing2.add(7.7);
+    bearing2.add(8.8);
+
+    List<List<Double>> bearings = new ArrayList<>();
+    bearings.add(null);
+    bearings.add(null);
+    bearings.add(bearing1);
+    bearings.add(bearing2);
+    bearings.add(null);
+
+    RouteOptions routeOptions = routeOptions()
+        .toBuilder()
+        .bearingsList(bearings)
+        .build();
+
+    assertEquals(bearings, routeOptions.bearingsList());
+    assertEquals(";;1.1,2.2;7.7,8.8", routeOptions.bearings());
+  }
+
+  @Test
+  public void approachesString() {
+    String approachesStr = ";" + APPROACH_CURB + ";" + ";" + APPROACH_UNRESTRICTED + ";";
+
+    RouteOptions routeOptions = routeOptions()
+        .toBuilder()
+        .approaches(approachesStr)
+        .build();
+
+    assertEquals(";" + APPROACH_CURB + ";" + ";" + APPROACH_UNRESTRICTED,
+        routeOptions.approaches());
+
+    List<String> approaches = routeOptions.approachesList();
+    assertEquals(4, approaches.size());
+    assertEquals(null, approaches.get(0));
+    assertEquals(APPROACH_CURB, approaches.get(1));
+    assertEquals(null, approaches.get(2));
+    assertEquals(APPROACH_UNRESTRICTED, approaches.get(3));
+  }
+
+  @Test
+  public void approachesList() {
+    List<String> approaches = new ArrayList<>();
+    approaches.add(APPROACH_CURB);
+    approaches.add(null);
+    approaches.add(null);
+    approaches.add(APPROACH_UNRESTRICTED);
+    approaches.add(APPROACH_UNRESTRICTED);
+    approaches.add(null);
+    approaches.add(null);
+    approaches.add(APPROACH_CURB);
+    approaches.add(null);
+
+    RouteOptions routeOptions = routeOptions()
+        .toBuilder()
+        .approachesList(approaches)
+        .build();
+
+    assertEquals(approaches, routeOptions.approachesList());
+    assertEquals("curb;;;unrestricted;unrestricted;;;curb", routeOptions.approaches());
+  }
+
+  @Test
+  public void waypointIndicesString() {
+    String indicesStr = "5;1;8;4;;;4;7;;;";
```

Comment:
This is not a valid waypoint indices output 👀 https://github.com/mapbox/mapbox-java/pull/1118#issuecomment-582521995 Am I missing something? Let's sync up internally if the example wasn't clear.

Reply:
@Guardiola31337 we check that we build a valid list even if a customer passes an invalid string. so, the list will be `(5, 1, 8, 4, null, null, 4, 7)`
and when we build a query string to make API request, we will get "5;1;8;4;;4;7"

Reply:
My points is that `"5;1;8;4;;4;7"` is not a valid input for the Directions API. It should be `"1;4;5;7;8"`  - in order, not duplicated, not trailing `""` (`;;`). Is that correct @danpaz?

Reply:
@Guardiola31337 is correct, there should not be null entries in the list. 

To expand a bit further: the purpose of the `waypoints` parameter is for the user to select which route coordinates should be treated as _break_ points along the route as opposed to _via_ points. Break points receive arrival and departure instructions and create multiple legs on the route. Via points are "silent" waypoints which do not receive arrival/departure instructions and don't break up a leg. The default behavior without waypoints is to treat every waypoint as a break point. 

For example, if the user supplied 5 lon,lat coordinate pairs:

```
0,0
1,1
2,2
3,3
4,4
```

and set waypoints=0;2;4, the result would be setting the first, third, and fifth coordinates as break points.

```
0,0  <-- treat as break point
1,1  <-- treat as via point
2,2  <-- treat as break point
3,3  <-- treat as via point
4,4  <-- treat as break point
```
The result would be a route with 2 legs instead of 4 legs. Hope this explanation helps.

Reply:
Is it fair to say that only `waypoints` cannot have `;;`  inside a string? We want to double check @danpaz as there are other parameters that accept trailing `""` (`";;"`).

Reply:
Yes waypoints is unique in that sense. 

It doesn't make sense to have an empty entry in waypoints, because each value is an index into the coordinates. Other parameters like `radiuses` can have empty entries because there's a default value to fall back to in that case.

Reply:
discussed. waypoint indices can't have empty values